src/OrleansProviders/Streams/Common/PooledCache/ObjectPool.cs

```diff
@@ -0,0 +1,44 @@
+﻿
+using System;
+using System.Collections.Generic;
+
+namespace Orleans.Providers.Streams.Common
+{
+    /// <summary>
+    /// Simple object pool that uses a stack to store available objects.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class ObjectPool<T> : IObjectPool<T>
+        where T : PooledResource<T>
+    {
+        private const int DefaultPoolCapacity = 1 << 10; // 1k
+        private readonly Stack<T> pool;
+        private readonly Func<IObjectPool<T>, T> factoryFunc;
+
+        public ObjectPool(Func<IObjectPool<T>, T> factoryFunc, int initialCapacity = DefaultPoolCapacity)
+        {
+            if (factoryFunc == null)
+            {
+                throw new ArgumentNullException("factoryFunc");
+            }
+            if (initialCapacity < 0)
+            {
+                throw new ArgumentOutOfRangeException("initialCapacity");
+            }
+            this.factoryFunc = factoryFunc;
+            pool = new Stack<T>(initialCapacity);
+        }
+
+        public virtual T Allocate()
+        {
+            return pool.Count != 0
+                ? pool.Pop()
+                : factoryFunc(this);
+        }
+
+        public void Free(T resource)
+        {
+            pool.Push(resource);
```

Comment:
OK, I talked with Jason in person, and it is in fact doing 2 layers of reducing GC pressure, and that was perf tested a lot in 343. That part might be fine, but I would still like to put a cap on the maximum number of PooledResources that would be stored in the stack waiting for re-allocation. Under normal load I wouldn't expect to have thousands of un-allocated objects standing by... I would prefer the GC to reclaim them... I would only expect a few of them at most (maybe 5 or 10?) waiting for reallocation. Still putting a cap that is bigger than that (50? 100?) I would consider safer.